core: error if chrome version does not support lcp metric

adamraine opened the pull request:
<!--
Thank you for submitting a pull request!
See CONTRIBUTING.MD for help in getting a change landed.
  https://github.com/GoogleChrome/lighthouse/blob/master/CONTRIBUTING.md
-->

**Summary**
<!-- What kind of change does this PR introduce? -->
<!-- Is this a bugfix, feature, refactoring, build related change, etc? -->
This change outputs a log message when the Chrome version is too low to support the LCP metric (<83).

<!-- Describe the need for this change -->

<!-- Link any documentation or information that would help understand this change -->

**Related Issues/PRs**
<!-- Provide any additional information we might need to understand the pull request -->
Issue: https://github.com/GoogleChrome/lighthouse/issues/10499

connorjclark commented on lighthouse-core/gather/gather-runner.js:
```diff
@@ -631,6 +631,11 @@ class GatherRunner {
       await GatherRunner.setupDriver(driver, options);
 
+      const milestone = (await driver.getBrowserVersion()).milestone;
+      if (milestone < 83) {
```
Ideally we could warn on "older than 2 Chrome versions" or something, but we don't have a way to know what the current stable version is (unless we were to extrapolate based on the release cadence, which, AFAICT, is reliable but feels hacky).

I wonder if we could possibly extend [Browser.getVersion](https://chromedevtools.github.io/devtools-protocol/tot/Browser/#method-getVersion) to return the release date. That may be the best solution, and as a first pass we can just hardcore `83` here.

paulirish replied:
That's a more general solution to older chrome's running against LH. I personally really like that we have specific milestone requirements scattered throughout.. Also makes it much easier for us to track the age of any particular assumption.

adamraine replied:
I will add a TODO for now

paulirish commented on lighthouse-core/gather/gather-runner.js:
```diff
@@ -631,6 +631,11 @@ class GatherRunner {
       await GatherRunner.setupDriver(driver, options);
 
+      const milestone = (await driver.getBrowserVersion()).milestone;
+      if (milestone < 83) {
```
Typically we place these checks closer to the intended use. (Someone using LH without the perf category doesn't have any problem)

@connorjclark so i'm thinking this could go in either the LCP audit itself or trace-processor. wdyt?


Also.. Oddly, our existing comment sez 78, but the bug reporter also had 78.
https://github.com/GoogleChrome/lighthouse/blob/b8bc05aa669913fdc5a7eb4f11ee1b9dc229d832/lighthouse-core/computed/metrics/largest-contentful-paint.js#L10

 So i suppose the 78 is wrong, but what about 79-82?

brendankenny replied:
> Typically we place these checks closer to the intended use. (Someone using LH without the perf category doesn't have any problem)

I was too late with my comment, agree with ☝️ 

> this could go in either the LCP audit itself or trace-processor

my vote is definitely in LCP as it keeps it local to what cares the most about whether or not there was an LCP trace event. `trace-processor` could in theory be looking at something that's not expected to have an LCP event and it's no big deal.

<hr>

But @paulirish I take it you want to abandon the "Add toplevelwarning about old Chrome" goal? I ask because you were the one to change the title of #10499 to that, but I would guess it was during a bug bash and we were all there :)

adamraine replied:
In the LCP audit it looks like we can fetch the chrome version with `artifacts.HostUserAgent` but that returns an unfriendly string like `Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) HeadlessChrome/83.0.4103.116 Safari/537.36`.  Is there a better way to fetch the chrome milestone in the audit?

paulirish replied:
@adamraine good call. right now that's the best we got. But on the bright side, we know it's a Chrome UA, so there's not much variability we have to deal with. So a pretty simple regex should do the job.

connorjclark replied:
```
'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) HeadlessChrome/83.0.4103.116 Safari/537.36'.match(/Chrome\/(\d+)/)[1]
```

fyi

brendankenny commented:
can someone summarize the conclusion of what we wanted in #10499? :)

If this is just for LCP, it makes sense to live in the LCP audit. The browser version check is a little more work, but not too bad.

#10499's title was also changed to just "Add toplevelwarning about old Chrome", in which case it does make sense to live in `gather-runner`, but that's a more difficult condition to track as @connorjclark outlines.

connorjclark commented:
There's also that CLS is missing important information in the trace until m84 https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-cli/test/smokehouse/test-definitions/perf/expectations.js#L178 , which lands in 3 weeks. I was hoping we'd have one place to say "we noticed that your Chrome is older than Stable, so some features may be missing". It's unfortunate this conflicts with our preference to generate the errors where they happen. Should we just be OK with generating multiple, specific errors? so for m82 you'd get two errors, one citing CLS and another LCP.

brendankenny commented:
> Should we just be OK with generating multiple, specific errors? so for m82 you'd get two errors, one citing CLS and another LCP.

If it's a too old error, yeah, there should only be one at the top level. A parameterized `NO_LCP` with a reason for it (in this case, too old) is another option to make clear what's happening instead of a top-level warning.

adamraine commented:
Does it still make sense to push the warning to `LighthouseRunWarnings` if the check is moved inside the LCP audit?

connorjclark commented:
yea. you'll find it on `passContext.LighthouseRunWarnings`

patrickhulce commented:
If it's moved into the audit you'll return it as part of the audit result.

See bootup time for example
https://github.com/GoogleChrome/lighthouse/blob/b8bc05aa669913fdc5a7eb4f11ee1b9dc229d832/lighthouse-core/audits/bootup-time.js#L207-L215

paulirish commented:
> can someone summarize the conclusion of what we wanted in #10499? :)

hah yeah. :) good call.

---------

stepping back..

We have options:

* **run-level warning**: could be general version check or multiple specific ones
* **audit-level warning** (not currently supported in the report ui for metrics, but doable)
* **audit-level error:** a la parameterized `NO_LCP` with a reason

Keeping the error specific to the audit is ideal, so someone doing a11y-only doesn't need to deal with our perf requirements. IMO this is more important than the downside of potentially showing multiple warnings.

So.. excluding the global version check... I think the NO_FCP_OLD_CHROME-ish case that motivated this is served best by an **audit error.** Basically we're being more specific than just plain NO_LCP (much like we want to do for NO_FCP).


@connorjclark @brendankenny work for you?

@adamraine sorry for switching things up on you but yeah we hadn't really sorted out this consensus yet.  :)

brendankenny commented:
> So.. excluding the global version check... I think the NO_FCP_OLD_CHROME-ish case that motivated this is served best by an **audit error.** Basically we're being more specific than just plain NO_LCP (much like we want to do for NO_FCP).

Sounds good to me! As you said, no audit-level warnings for the perf metrics, but the audit-level error message is shown prominently in place.

Looking at the original error case in #10499, had the `NO_LCP` warning here

<img width="450" alt="Example LCP error in the Lighthouse report" src="https://user-images.githubusercontent.com/316891/85738191-d49f9300-b6cd-11ea-895d-9268e7f953ca.png">

instead mentioned the version of Chrome being too old, it would have helped the user find the problem immediately, so I think this will work well (for this and for a similar CLS error @connorjclark mentioned above).

patrickhulce commented:
So does this mean we're not implementing what #10499 was reopened for, a toplevel warning when the Chrome version is older than whatever feature set we are assuming in Lighthouse is available? :)

I'm fine with throwing a more specific LCP error too, but I want to make sure we don't *just* implement that and consider #10499 resolved.

CLS error is tricky because no events is also a valid case. Is the plan now to sprinkle `if (artifacts.ChromeMilestone < XX) throw` into each audit that requires some chrome version whether it would actually cause the audit to fail or not?

adamraine commented:
I updated the error to emit inside the LCP audit and also changed the minimum version to 79.

According to this doc chrome 79 is the minimum version for LCP but could someone please confirm.
https://chromium.googlesource.com/chromium/src/+/master/docs/speed/metrics_changelog/lcp.md

The current solution just replaces all NO_LCP errors caught in the audit function with "Version too low" errors but I'm not sure this is the best solution if NO_LCP could emit for reasons other than outdated chrome.

adamraine commented:
> you'll probably want a new one that passes and an old one that fails the version check

Wouldn't the existing test cover the case where the LCP passes the version check?

brendankenny commented:
> > you'll probably want a new one that passes and an old one that fails the version check
> 
> Wouldn't the existing test cover the case where the LCP passes the version check?

The version check won't happen if there's not a `NO_LCP` error thrown from `ComputedLcp.request()`, so it will have to be two tests of a trace without a NO_LCP event to cover both sides of the version check (one using a new UA string like Chrome 84 (or whatever) and an older Chrome 74 (or whatever)).

brendankenny commented on lighthouse-core/test/audits/metrics/largest-contentful-paint-test.js:
```diff
@@ -27,20 +31,53 @@ function generateContext({throttlingMethod}) {
+      trace,
+      devtoolsLog,
+      TestedAsMobileDevice: true,
+      HostUserAgent: 'Chrome/78.0.3904.70',
```
probably should use a full UA string, just so it better matches what the real input will be.

e.g if we ever move to a centralized UA string parser function and it used a regex that does multiple capture groups on the OS and whatever else so would only match against the whole string, ideally these tests would keep passing without having to be updated.

```suggestion
      HostUserAgent: 'Mozilla/5.0 (Linux; Android 6.0.1; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.70 Mobile Safari/537.36 Chrome-Lighthouse',
```
(and for the other `HostUserAgent`s below)

brendankenny commented:
> I re-ran the tests to hopefully unflake

looks all good! Over to you @connorjclark
